tests/system/Database/Live/UpsertTest.php

```diff
@@ -0,0 +1,613 @@
+                    "name" = "excluded"."name"
+                    SQL;
+                break;
+
+            case 'SQLite3':
+                $expected = <<<'SQL'
+                    INSERT INTO `db_user` (`country`, `email`, `name`)
+                    VALUES ('Iran','ahmadinejad@world.com','Ahmadinejad')
+                    ON CONFLICT(`email`)
+                    DO UPDATE SET
+                    `country` = `excluded`.`country`,
+                    `name` = `excluded`.`name`
+                    SQL;
+                break;
+
+            case 'SQLSRV':
+                $expected = <<<'SQL'
+                    MERGE INTO "test"."dbo"."db_user"
+                    USING (
+                    VALUES ('Iran','ahmadinejad@world.com','Ahmadinejad')
+                    ) "_upsert" ("country", "email", "name")
+                    ON ("test"."dbo"."db_user"."email" = "_upsert"."email")
+                    WHEN MATCHED THEN UPDATE SET
+                    "country" = "_upsert"."country",
+                    "name" = "_upsert"."name"
+                    WHEN NOT MATCHED THEN INSERT ("country", "email", "name")
+                    VALUES ("_upsert"."country", "_upsert"."email", "_upsert"."name");
+                    SQL;
+                break;
+
+            case 'OCI8':
+                $expected = <<<'SQL'
+                    MERGE INTO "db_user"
+                    USING (
+                    SELECT 'Iran' "country", 'ahmadinejad@world.com' "email", 'Ahmadinejad' "name" FROM DUAL
+                    ) "_upsert"
+                    ON ("db_user"."email" = "_upsert"."email")
+                    WHEN MATCHED THEN UPDATE SET
+                    "country" = "_upsert"."country",
+                    "name" = "_upsert"."name"
+                    WHEN NOT MATCHED THEN INSERT ("country", "email", "name")
+                    VALUES  ("_upsert"."country", "_upsert"."email", "_upsert"."name")
+                    SQL;
+                break;
+
+            default:
+                $expected = false;
+                break;
+        }
+
+        $userData = [
+            'email'   => 'ahmadinejad@world.com',
+            'name'    => 'Ahmadinejad',
+            'country' => 'Iran',
+        ];
+
+        $this->assertSame($expected, $this->db->table('user')
+            ->set($userData)
+            ->getCompiledUpsert());
+    }
+
+    public function testGetCompiledUpsertBatch()
+    {
+        $userData = [
+            [
+                'email'   => 'ahmadinejad@example.com',
+                'name'    => 'Ahmadinejad',
+                'country' => 'Iran',
+            ],
+            [
+                'email'   => 'pedro@example.com',
+                'name'    => 'Pedro',
+                'country' => 'El Salvador',
+            ],
+        ];
+
+        $sql = $this->db->table('user')
+            ->setData($userData)
+            ->onConstraint('email')
+            ->getCompiledUpsert();
```

Comment:
It seems something wrong. The `$sql` value  is like this:
```
string (254) "INSERT INTO `db_user` (0, 1, 2)
VALUES ('\'Iran\'','\'ahmadinejad@example.com\'','\'Ahmadinejad\''), ('\'El Salvador\'','\'pedro@example.com\'','\'Pedro\'')
ON DUPLICATE KEY UPDATE
`db_user`.0 = VALUES(0),
`db_user`.1 = VALUES(1),
`db_user`.2 = VALUES(2)"
```

```
string (244) "INSERT INTO `db_user` (0, 1, 2)
VALUES ('''Iran''','''ahmadinejad@example.com''','''Ahmadinejad'''), ('''El Salvador''','''pedro@example.com''','''Pedro''')
ON CONFLICT(`email`)
DO UPDATE SET
0 = `excluded`.0,
1 = `excluded`.1,
2 = `excluded`.2"
```

Reply:
Fixed this. In my last fix I had not had the proper conditions evaluate before running setData().

These issues stem from allowing the use of set() and setData() to both be used with upsert(). For some reason set() uses binds except when not escaping. So if QBSet has been created with set() I have to get the data and set it with setData(). This way it is an array of arrays/objects. The error was caused because setData() removes the associated array and separates the columns to QBKeys.